Add internal compaction API for Secondary instance

jay-zhuang opened the pull request:
Add compaction API for secondary instance, which compact the files to a secondary DB path without installing to the LSM tree.
The API will be used to remote compaction.

Test plan: `make check`

ajkr commented on db/db_impl/db_impl_secondary.cc:
```diff
@@ -17,8 +17,9 @@ namespace ROCKSDB_NAMESPACE {
-    : DBImpl(db_options, dbname) {
+                                 const std::string& dbname,
+                                 const std::string& secondary_path)
+    : DBImpl(db_options, dbname), secondary_path_(secondary_path) {
```
(unrelated to this PR) I think `DBImpl` should be passed `read_only=true`. That's a new flag introduced in #8142 to prevent `SanitizeOptions()` from creating an info logger when none is provided (i.e., `info_log == nullptr`). I'd guess in a secondary DB, a `SanitizeOptions()`-created info logger could conflict with the primary DB info logger. It looks like `OpenAsSecondary()` tries to avoid opening the DB with option `info_log == nullptr` -- except when  L616 is executed.

jay-zhuang replied:
Thanks for pointing it out, I need look into it to understand the problem, filed an issue to myself: #8193

ajkr commented on db/db_impl/db_impl_secondary.cc:
```diff
@@ -663,6 +664,103 @@ Status DB::OpenAsSecondary(
+    return s;
+  }
+
+  LogBuffer log_buffer(InfoLogLevel::INFO_LEVEL,
```
Does this need to be flushed somewhere? I don't know why `CompactionJob` requires this, but it feels surprising if the client is expected to own it but doesn't need to do anything with it after the compaction finishes.

jay-zhuang replied:
Isn't it flushed in `CompactionJob::Run()`: https://github.com/facebook/rocksdb/blob/d89483098ff4a77bd0ef12454a64554805984cc5/db/compaction/compaction_job.cc#L586

ajkr reviewed (approved):
LGTM, good to see `CompactionJob` can support local execution and remote execution. What will the primary side look like during remote execution? Do we also create a `CompactionJob` there and invoke some subset of its operations?

riversand963 commented on db/db_impl/db_impl_secondary.cc:
```diff
@@ -17,8 +17,9 @@ namespace ROCKSDB_NAMESPACE {
-                                 const std::string& dbname)
-    : DBImpl(db_options, dbname) {
+                                 const std::string& dbname,
+                                 const std::string& secondary_path)
```
Nit: since we are going to store `secondary_path_` as a copy, we can make `secondary_path` of type `std::string`, and do `secondary_path_(std::move(secondary_path))`.

riversand963 commented on db/db_impl/db_impl_secondary.cc:
```diff
@@ -663,6 +664,103 @@ Status DB::OpenAsSecondary(
+
+  // TODO: there's only 1 job could be run per DB secondary instance, add
+  // validation for that.
+  const int job_id = 1;
```
Can we still use `next_job_id_`?
> TODO: there's only 1 job could be run per DB secondary instance

I am not sure whether you mean "only 1 job ID" or "only 1 job", because they do not seem to contradict with each other.

riversand963 reviewed (approved):
LGTM. Thanks @jay-zhuang for the PR

ajkr commented:
> What will the primary side look like during remote execution? Do we also create a CompactionJob there and invoke some subset of its operations?

Overall I'm finding it difficult to wrap my head around the different modes of `CompactionJob` (on primary w/ local execution, on primary w/ remote execution, and remote execution). Unless I'm missing something, it seems both the `CompactionJob` client and the `CompactionJob` itself need to behave differently in these different modes. The client needs to call the right subset of `CompactionJob` functions according to the mode, and the functions themselves need to behave differently according to the mode.

Have we tried prototyping `ProcessKeyValueCompaction()` as the unit of work that can be relocated to a remote worker? I still suspect keeping control logic all on the primary can help. I recall we found execution logic residing outside of `ProcessKeyValueCompaction()` (i.e., fsync'ing parent directory and verifying output tables). After thinking through it more, it makes sense to me to do the table verification in `ProcessKeyValueCompaction()`, and that function can fsync the secondary_path directory too when run remotely.

jay-zhuang commented on db/compaction/compaction_job.cc:
```diff
@@ -899,6 +910,35 @@ Status CompactionJob::Install(const MutableCFOptions& mutable_cf_options) {
+  for (auto& output_file : subcompaction_state.outputs) {
+    auto& meta = output_file.meta;
+    compaction_result->output_files.emplace_back(
+        MakeTableFileName(".", meta.fd.GetNumber()), meta.fd.smallest_seqno,
```
I guess we don't need the "." here.

jay-zhuang commented on db/compaction/compaction_job.h:
```diff
@@ -57,6 +57,81 @@ class Version;
+struct CompactionServiceInput {
+  std::string column_family_name;
+
+  DBOptions db_options;
```
It has column_family_name and `ColumnFamilyOptions`. I'll combine them with `ColumnFamilyDescriptor`.
I don't think we need vector here, as compaction is only for 1 column family.

jay-zhuang commented:
> What will the primary side look like during remote execution? Do we also create a CompactionJob there and invoke some subset of its operations?

Yes, the CompactionJob will be created and then `Prepare()` -> `Run()` -> `Install()`. The remote_compaction part will be handled inside of `Run()`, instead of calling `ProcessKeyValueCompaction()`, it will call the CompactionService callback.

> Overall I'm finding it difficult to wrap my head around the different modes of `CompactionJob` (on primary w/ local execution, on primary w/ remote execution, and remote execution). Unless I'm missing something, it seems both the `CompactionJob` client and the `CompactionJob` itself need to behave differently in these different modes. The client needs to call the right subset of `CompactionJob` functions according to the mode, and the functions themselves need to behave differently according to the mode.

Yeah, I agree, I updated the patch to have a `CompactionServiceCompactionJob` for remote side, which differentiate itself from primary's `CompactionJob` and also group things together.

> Have we tried prototyping `ProcessKeyValueCompaction()` as the unit of work that can be relocated to a remote worker? I still suspect keeping control logic all on the primary can help. I recall we found execution logic residing outside of `ProcessKeyValueCompaction()` (i.e., fsync'ing parent directory and verifying output tables). After thinking through it more, it makes sense to me to do the table verification in `ProcessKeyValueCompaction()`, and that function can fsync the secondary_path directory too when run remotely.

I updated it to have `CompactionServiceCompactionJob` to call `ProcessKeyValueCompaction()` directly instead of using the same `Prepare()` and `Run()`. It does make it cleaner.

ajkr commented on db/compaction/compaction_job.h:
```diff
@@ -194,19 +202,130 @@ class CompactionJob {
+  uint64_t approx_size;
+  uint64_t bytes_read;
+  uint64_t bytes_written;
+  CompactionJobStats stats;
```
Is it used?

ajkr reviewed (approved):
LGTM! Nice separation of the remote compaction work into a dedicated class with a more targeted API.
